fastseq/optimizer/fairseq/el_attention_optimizer.py

```diff
@@ -0,0 +1,1459 @@
+        x = x.transpose(0, 1)
+
+        self_attn_padding_mask = None
+        if (self.cross_self_attention or
+                prev_output_tokens.eq(self.padding_idx).any()):
+            self_attn_padding_mask = prev_output_tokens.eq(self.padding_idx)
+
+        # decoder layers
+        attn = None
+        inner_states = [x]
+        for idx, layer in enumerate(self.layers):
+            encoder_state = None
+            if encoder_out is not None:
+                if self.layer_wise_attention:
+                    encoder_state = encoder_out.encoder_states[idx]
+                else:
+                    encoder_state = encoder_out.encoder_out
+                    encoder_out_v = encoder_out.encoder_out_v
+
+            if incremental_state is None and not full_context_alignment:
+                self_attn_mask = self.buffered_future_mask(x)
+            else:
+                self_attn_mask = None
+
+            # add LayerDrop (see https://arxiv.org/abs/1909.11556 for description)
+            dropout_probability = random.uniform(0, 1)
+            if not self.training or (dropout_probability
+                    > self.decoder_layerdrop):
+                x, layer_attn = layer(
+                    x,
+                    encoder_state,
+                    encoder_out_v,
+                    (encoder_out.encoder_padding_mask
+                    if encoder_out is not None else None),
+                    incremental_state,
+                    self_attn_mask=self_attn_mask,
+                    self_attn_padding_mask=self_attn_padding_mask,
+                    need_attn=(idx == alignment_layer),
+                    need_head_weights=(idx == alignment_layer),
+                )
+                inner_states.append(x)
+                if layer_attn is not None and idx == alignment_layer:
+                    attn = layer_attn.float()
+
+        if attn is not None:
+            if alignment_heads is not None:
+                attn = attn[:alignment_heads]
+
+            # average probabilities over heads
+            attn = attn.mean(dim=0)
+
+        if self.layer_norm:
+            x = self.layer_norm(x)
+
+        # T x B x C -> B x T x C
+        x = x.transpose(0, 1)
+
+        if self.project_out_dim is not None:
+            x = self.project_out_dim(x)
+
+        return x, {'attn': attn, 'inner_states': inner_states}
+
+
+@replace(FairseqEncoderDecoderModel, use_el_attn)
+class FairseqEncoderDecoderModelV2(FairseqEncoderDecoderModel):
+    """class for encoder-decoder models.
+    Args:
+        encoder (FairseqEncoder): the encoder
+        decoder (FairseqDecoder): the decoder
+    """
+
+    def transpose_enc_dec_kv_proj (self):
+        for i in range (0, len(self.decoder.layers)):
+            self.num_heads = self.decoder.layers[i].encoder_attn.num_heads
+            self.head_dim = self.decoder.layers[i].encoder_attn.head_dim
+
+
+            self.decoder.layers[i].encoder_attn.k_proj_weight_t = (
```

Comment:
Add a TODO. Keep the original k/v weight in CPU or remove them to avoid waste GPU memory.

Reply:
deleted original weights after transposing.